src/amo/components/PermissionsCard/permissions.js

```diff
@@ -0,0 +1,238 @@
+    const platform = userAgentOSToPlatform[agentOsName];
+
+    if (!platform) {
+      log.error(oneLine`No platform exists for user agent OS 
+        "${agentOsName || '<undefined agentOsName>'}"`);
+      return [];
+    }
+
+    const file = addon.platformFiles[platform] || addon.platformFiles[OS_ALL];
+
+    if (!file) {
+      log.debug(oneLine`No file exists for platform "${agentOsName}"
+        (mapped to "${platform}"); platform files:`, addon.platformFiles);
+      return [];
+    }
+    return file.permissions || [];
+  }
+
+  // Classify a permission as a host permission or a regular permission.
+  classifyPermission(
+    permission: string,
+  ): {| type: 'permissions' | 'hosts', value: string |} {
+    const match = /^(\w+)(?:\.(\w+)(?:\.\w+)*)?$/.exec(permission);
+    let result = { type: 'permissions', value: permission };
+    if (!match) {
+      result = { type: 'hosts', value: permission };
+    }
+    return result;
+  }
+
+  // Format and sequence all the Permission components.
+  formatPermissions(addonPermissions: Array<string>): Array<Object> {
+    const permissionsToDisplay = [];
+    const permissions = { hosts: [], permissions: [] };
+
+    // First, categorize them into host permissions and regular permissions.
+    for (const permission of addonPermissions) {
+      const { type, value } = this.classifyPermission(permission);
+      permissions[type].push(value);
+    }
+
+    // Classify the host permissions.
+    let allUrls = false;
+    const wildcards = [];
+    const sites = [];
+    for (const permission of permissions.hosts) {
+      if (permission === '<all_urls>') {
+        allUrls = true;
+        break;
+      }
+      if (permission.startsWith('moz-extension:')) {
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+      const match = /^[a-z*]+:\/\/([^/]+)\//.exec(permission);
+      if (!match) {
+        log.debug(`Host permission string "${permission}" appears to be invalid.`);
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+      if (match[1] === '*') {
+        allUrls = true;
+      } else if (match[1].startsWith('*.')) {
+        wildcards.push(match[1].slice(2));
+      } else {
+        sites.push(match[1]);
+      }
+    }
+
+    // Format the host permissions.  If we have a wildcard for all urls,
+    // a single string will suffice.  Otherwise, show domain wildcards
+    // first, then individual host permissions.
+
+    if (allUrls) {
+      permissionsToDisplay.push(
+        <Permission
+          className="hostPermission"
+          description={this.permissionStrings.allUrls}
+          key="allUrls"
+        />
+      );
+    } else {
+      // Formats a list of host permissions.  If we have 4 or fewer, display
+      // them all, otherwise display the first 3 followed by an item that
+      // says "...plus N others".
+      const formatHostPermissions = (list, itemKey, moreKey) => {
```

Comment:
I don't want to hold up this patch so here's a comment for future reference -- it's not a change request.

React is really good at encapsulating formatting code and encouraging you to use composition to put all formatters together. The short lived inline functions here suggest to me that the code should be split into more components. 

In this case, the logic for rendering host/site permissions is quite complex and is also used in more than one place. It might work better as a separate `<HostPermissions />` component that would be responsible for handling `<all_urls>` and wildcards, rendering the right number of `<Permission>` components, and adding the "...and 3 more..." permission. Ultimately, it would need to return a list of `<li>` tags without a parent element which would require a [fragment](https://reactjs.org/docs/fragments.html).

Again, I don't want to hold up this patch but it could be something to work on in a low priority follow-up issue.

Reply:
This is an interesting idea, and one I'd like to pursue, but yes, doing so in a follow-up seems to make sense so we can land this feature sooner rather than later. I opened #4527 to track it.